rep-2011.rst

```diff
@@ -0,0 +1,1421 @@
+                                  ▼ ▼ ▼
+    ┌──────────────────────┐                  ┌─────────────────────────┐
+    │Publisher<Message@ABC>│                  │Subscription<Message@XYZ>│
+    └─┬────────────────────┘                  └─────────────────────────┘
+      │                                                          ▲
+      │             ┌─────────────────────────────────┐          │
+      ✓ /topic/ABC  │ Transfer Functions for ABC->XYZ │   /topic ✓
+      │             │                                 │          │
+      │  ┌──────────┴──────────────┐   ┌──────────────┴───────┐  │
+      └─►│Subscription<Message@ABC>│   │Publisher<Message@XYZ>├──┘
+         └──────────┬──────────────┘   └──────────────┬───────┘
+                    │                                 │
+                    └─────────────────────────────────┘
+
+Once the set of necessary transfer functions has been identified, the ROS graph can be changed to have one side of the topic be remapped onto a new topic name which indicates it is of a different version that what is desired, and then the transfer function can be run as a component node which subscribes to one version of the message, performs the conversion using the chain of transfer functions, and then publishes the other version of the message.
+Tools will assist the user in making these remappings and running the necessary component nodes with the appropriate configurations, either from their launch file or from the command line.
+
+.. TODO::
+
+    discuss the implications for large messages and the possibility of having the transfer functions be colocated with either the publisher or subscription more directly than with component nodes and remapping.
+
+Once the mismatched messages are flowing through the transfer functions, communication should be possible and neither the publishing side nor the subscribing side have any specific knowledge of the conversions taking place or that any conversions are necessary.
+
+.. TODO::
+
+    Extend conceptual overview to describe how this will work with Services and Actions.
+    Services, since they are not sensitive to the many-to-many (many publisher) issue, unlike Topics, and because they do not have as many QoS settings that apply to them, they can probably have transfer functions that are plugins, rather than separate component nodes that repeat the service call, like the ros1_bridge.
+    Actions will be a combination of topics and services, but will have other considerations in the tooling.
+
+In order to support this vision, three missing features will need to be added into ROS 2 (which were also mentioned in the introduction):
+
+- enforcing type compatibility by version
+- providing access to the complete type description of types being used
+- making it possible to publish and subscribe to topics using just the type description
+
+These features are described in the following sections.
+
+Type Version Enforcement
```

Comment:
It is more like detection as feature? user must be able to know if the type version mismatches or not, then it is up to user application to enforce and transfer the version?